src/utilities/applyRequiredStatus.ts

```diff
@@ -0,0 +1,109 @@
+import { GraphQLError } from '../error/GraphQLError';
+
+import type {
+  ListNullabilityNode,
+  NullabilityDesignatorNode,
+} from '../language/ast';
+import type { ASTReducer } from '../language/visitor';
+import { visit } from '../language/visitor';
+
+import type { GraphQLOutputType } from '../type/definition';
+import {
+  assertListType,
+  getNullableType,
+  GraphQLList,
+  GraphQLNonNull,
+  isListType,
+  isNonNullType,
+} from '../type/definition';
+
+/**
+ * Implements the "Accounting For Client Controlled Nullability Designators"
+ * section of the spec. In particular, this function figures out the true return
+ * type of a field by taking into account both the nullability listed in the
+ * schema, and the nullability providing by an operation.
+ */
+export function applyRequiredStatus(
+  type: GraphQLOutputType,
+  nullabilityNode?: ListNullabilityNode | NullabilityDesignatorNode,
+): GraphQLOutputType {
+  const typeStack: [GraphQLOutputType] = [type];
+
+  while (isListType(getNullableType(typeStack[typeStack.length - 1]))) {
+    const list = assertListType(
+      getNullableType(typeStack[typeStack.length - 1]),
+    );
+    const elementType = list.ofType as GraphQLOutputType;
+    typeStack.push(elementType);
+  }
+
```

Comment:
Only populate typeStack after we've confirmed that `nullabilityNode` exists to improve performance